seed/seed20.go

```diff
@@ -135,47 +175,155 @@ func (s *seed20) Copy(seedDir string, label string, tm timings.Measurer) (err er
 	span := tm.StartSpan("copy-recovery-system", fmt.Sprintf("copy recovery system from %s to %s", srcSystemDir, destSystemDir))
 	defer span.Stop()
 
-	// copy all files (including unasserted snaps) from the seed to the
-	// destination
-	err = filepath.Walk(srcSystemDir, func(path string, info fs.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
-
-		destPath := filepath.Join(destSeedDir, "systems", label, strings.TrimPrefix(path, srcSystemDir))
-		if info.IsDir() {
-			return os.Mkdir(destPath, info.Mode())
-		}
-
-		return osutil.CopyFile(path, destPath, osutil.CopyFlagDefault)
-	})
+	entries, err := os.ReadDir(srcSystemDir)
 	if err != nil {
 		return err
 	}
 
-	destAssertedSnapDir := filepath.Join(destSeedDir, "snaps")
+	// copy all of the top-level files and directories from the seed. we'll copy
+	// the rest manually
+	for _, e := range entries {
+		if e.IsDir() {
+			if err := os.Mkdir(filepath.Join(destSystemDir, e.Name()), 0755); err != nil {
+				return err
+			}
+		} else {
+			if err := osutil.CopyFile(
+				filepath.Join(srcSystemDir, e.Name()),
+				filepath.Join(destSystemDir, e.Name()),
+				osutil.CopyFlagDefault,
+			); err != nil {
+				return err
+			}
+		}
+	}
 
-	if err := os.MkdirAll(destAssertedSnapDir, 0755); err != nil {
+	destAssertedSnapDir := filepath.Join(destSeedDir, "snaps")
+	if err := os.Mkdir(destAssertedSnapDir, 0755); err != nil {
 		return err
 	}
 
-	// copy the asserted snaps that the seed needs
+	destUnassertedSnapDir := filepath.Join(destSystemDir, "snaps")
+
+	// while copying the snaps, we also collect the assertions and the fields in
+	// the aux-info.json file that we need to write
+	var snapAssertions []asserts.Assertion
+	auxInfo := make(map[string]*internal.AuxInfo20)
+
+	// copy the snaps and components that the seed needs
 	for _, sn := range s.snaps {
-		// unasserted snaps are already copied above, skip them
-		if sn.ID() == "" {
+		// if we're not copying the snap, then we also don't need to copy the
```

Comment:
nitpick: I wonder how easy would be to have the content of this loop in a separate method to make Copy smaller.

Reply:
Did this in 80883c8d889e1bd42188079455265fcce358c951.